File: leetcode/pyleetcode/pyleetcode/easy/palindrome_number.py
"""9. Palindrome Number.

https://leetcode.com/problems/palindrome-number/
"""


class Solution:

    def isPalindrome(self, x: int) -> bool:  # noqa
        if x < 0 or (x != 0 and x % 10 == 0):
            return False
        invert = 0
        while x > invert:
            invert = invert * 10 + x % 10
            x //= 10

        return x == invert or x == invert // 10

File: leetcode/rsleetcode/src/easy/palindrome_number.rs
//! 9. Palindrome Number
//!
//! https://leetcode.com/problems/palindrome-number/

pub struct Solution {
}

impl Solution {
    pub fn is_palindrome(x: i32) -> bool {
        if x.is_negative() || (x != 0 && x % 10 == 0) {
            return false
        }
        let mut y = x;
        let mut invert = 0;
        while y > invert {
            invert = 10 * invert + y % 10;
            y /= 10;
        }

        invert == y || invert / 10 == y
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_1() {
        assert!(Solution::is_palindrome(121))
    }

    #[test]
    fn test_2() {
        assert!(!Solution::is_palindrome(-121))
    }

    #[test]
    fn test_3() {
        assert!(!Solution::is_palindrome(10))
    }

    #[test]
    fn test_4() {
        assert!(Solution::is_palindrome(0))
    }
}
